### PR TITLE
Resend the push installation log with authentication after successful login

### PR DESCRIPTION
**Pull request by annakocheshkova**
* [ ] Has `CHANGELOG.md` been updated?
* [X] Are tests passing locally?
* [X] Are the files formatted correctly?
* [X] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [X] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Resend the Push installation log with the new authentication after a successful login.

### CODE REVIEW

File: sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java [hunk 1]
```diff
@@ -293,6 +300,7 @@ public synchronized void onStarted(@NonNull Context context, @NonNull Channel ch
             AppCenterLog.debug(LOG_TAG, "Disabling Firebase analytics collection by default.");
             setFirebaseAnalyticsEnabled(context, false);
         }
+        AuthTokenContext.getInstance(mContext).addListener(this);
```
Comment (MatkovIvan) [inline sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java]: It will work even if push is disabled

Comment (annakocheshkova) [reply sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java]: :+1:

File: sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java [hunk 2]
```diff
@@ -298,6 +303,15 @@ protected int getTriggerCount() {
     @Override
     public synchronized void onStarted(@NonNull Context context, @NonNull Channel channel, String appSecret, String transmissionTargetToken, boolean startedFromApp) {
         mContext = context;
+        mAuthListener = new AbstractTokenContextListener() {
+            @Override
+            public synchronized void onNewUser(String authToken) {
+                if (mLatestPushToken != null) {
+                    enqueuePushInstallationLog(mLatestPushToken);
+                }
+
```
Comment (MatkovIvan) [inline sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java]: Extra line

File: sdk/appcenter-push/src/test/java/com/microsoft/appcenter/push/PushTest.java [hunk 3]
```diff
@@ -742,18 +742,49 @@ public void verifyEnqueueCalledOnNewAuthToken() {
...
+        start(push, channel);
+        push.onTokenRefresh("push-token");
+        String mockToken = UUIDUtils.randomUUID().toString();
+        String mockHomeId = UUIDUtils.randomUUID().toString();
+        AuthTokenContext.getInstance().setAuthToken(mockToken, mockHomeId);
+        AuthTokenContext.getInstance().setAuthToken(mockToken, "new-id");
+        verify(channel, times(3)).enqueue(any(com.microsoft.appcenter.ingestion.models.Log.class), anyString(), anyInt());
+    }
+
+
```
Comment (MatkovIvan) [inline sdk/appcenter-push/src/test/java/com/microsoft/appcenter/push/PushTest.java]: Extra line

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java [hunk 4]
```diff
@@ -156,7 +157,12 @@ public DefaultChannel(@NonNull Context context, String appSecret, @NonNull LogSe
         mEnabled = true;
         mAuthTokenContext = AuthTokenContext.getInstance();
         mIngestion.setAuthToken(mAuthTokenContext.getAuthToken());
-        mAuthTokenContext.addListener(this);
+        mAuthTokenContext.addListener(new AbstractTokenContextListener() {
+            @Override
```
Comment (MatkovIvan) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java]: Missing empty line

File: sdk/appcenter-push/src/test/java/com/microsoft/appcenter/push/PushTest.java [hunk 5]
```diff
@@ -730,4 +733,70 @@ private static Intent createPushIntent(String title, String message, final Map<S
         }
         return pushIntentMock;
     }
-}
+
+    @Test
+    public void verifyEnqueueCalledOnNewAuthToken() {
```
Comment (jwallra) [inline sdk/appcenter-push/src/test/java/com/microsoft/appcenter/push/PushTest.java]: Move the new tests above the private helper method

File: sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java [hunk 6]
```diff
@@ -288,6 +303,14 @@ protected int getTriggerCount() {
     @Override
     public synchronized void onStarted(@NonNull Context context, @NonNull Channel channel, String appSecret, String transmissionTargetToken, boolean startedFromApp) {
         mContext = context;
+        mAuthListener = new AbstractTokenContextListener() {
+            @Override
```
Comment (guperrot) [inline sdk/appcenter-push/src/main/java/com/microsoft/appcenter/push/Push.java]: Please add a blank line before annotations.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.microsoft.appcenter.utils.context;
+
+/**
+ * Empty implementation to make callbacks optional.
+ */
+public class AbstractTokenContextListener implements  AuthTokenContext.Listener {
+
+    @Override
+    public void onNewAuthToken(String authToken) {
+
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java]: Remove empty line for empty implementation

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java [hunk 8]
```diff
@@ -88,15 +88,19 @@ public synchronized String getAuthToken() {
...
      */
-    public void setAuthToken(String authToken) {
+    public void setAuthToken(String authToken, String homeAccountId) {
         mAuthToken = authToken;
-        //mLastHomeAccountId = homeAccountId;
 
         /* Call listeners so that they can react on new token. */
         for (Listener listener : mListeners) {
             listener.onNewAuthToken(authToken);
+            if (mLastHomeAccountId == null || !mLastHomeAccountId.equals(homeAccountId)) {
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java]: is the threading ok here or do we need synchronized method?

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java [hunk 9]
```diff
@@ -40,7 +41,7 @@
 
 import static com.microsoft.appcenter.AppCenter.LOG_TAG;
 
-public class DefaultChannel implements Channel, AuthTokenContext.Listener {
+public class DefaultChannel implements Channel{
```
Comment (dhei) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java]: nit: missing a space before `{`

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package com.microsoft.appcenter.utils.context;
+
+/**
+ * Empty implementation to make callbacks optional.
+ */
+public class AbstractTokenContextListener implements  AuthTokenContext.Listener {
```
Comment (jaelim-ms) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java]: Rename the class to `DefaultTokenContextListner` This isn't an abstract class.

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java]: or make it abstract to match other classes

Comment (MatkovIvan) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java]:
```suggestion
public abstract class AbstractTokenContextListener implements AuthTokenContext.Listener {
```

Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AbstractTokenContextListener.java]: (also extra space is removed)

**Comment by annakocheshkova** [comment]
@guperrot was a simple merge conflict. Now resolved.

**Comment by annakocheshkova** [comment]
@guperrot several lines got duplicated due to merge. All working now.

**Comment by annakocheshkova** [comment]
@guperrot what you are saying makes sense except for one minor thing.
If we disable `AppCenter` module, that means we are disabling everything and by logic we *should not* re-send any `Push` logs (cause `Push` gets disabled, too, and usually *before* `Identity`).
What I'm suggesting is to re-send anonymous log only if `Identity` module gets disabled. That would mean that `Push` is still working and actually needs re-sending anonymous logs.

File: sdk/appcenter-push/src/test/java/com/microsoft/appcenter/push/PushTest.java [hunk 11]
```diff
@@ -734,6 +734,19 @@ public void verifyEnqueueCalledOnNewAuthToken() {
         verify(channel, times(2)).enqueue(any(com.microsoft.appcenter.ingestion.models.Log.class), anyString(), anyInt());
     }
 
+    @Test
+    public void verifyEnqueueCalledAnonimouslyOnClearToken() {
```
Comment (dhei) [inline sdk/appcenter-push/src/test/java/com/microsoft/appcenter/push/PushTest.java]: typo: Anonymously

**Pull request merged**